bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/EntryLogTest.java

```diff
@@ -391,4 +400,137 @@ public void testGetEntryLogsSet() throws Exception {
 
         assertEquals(Sets.newHashSet(0L, 1L, 2L, 3L), logger.getEntryLogsSet());
     }
+
+    class LedgerStorageWriteTask implements Callable<Boolean> {
+        long ledgerId;
+        int entryId;
+        LedgerStorage ledgerStorage;
+
+        LedgerStorageWriteTask(long ledgerId, int entryId, LedgerStorage ledgerStorage) {
+            this.ledgerId = ledgerId;
+            this.entryId = entryId;
+            this.ledgerStorage = ledgerStorage;
+        }
+
+        @Override
+        public Boolean call() {
+            try {
+                ledgerStorage.addEntry(generateEntry(ledgerId, entryId));
+            } catch (IOException e) {
+                LOG.error("Got Exception for AddEntry call. LedgerId: " + ledgerId + " entryId: " + entryId, e);
+                return false;
+            }
+            return true;
+        }
+    }
+
+    class LedgerStorageReadTask implements Callable<Boolean> {
+        long ledgerId;
+        int entryId;
+        LedgerStorage ledgerStorage;
+
+        LedgerStorageReadTask(long ledgerId, int entryId, LedgerStorage ledgerStorage) {
+            this.ledgerId = ledgerId;
+            this.entryId = entryId;
+            this.ledgerStorage = ledgerStorage;
+        }
+
+        @Override
+        public Boolean call() {
```

Comment:
Instead of returning false on errors, you could throw Exception, and have that carry the context of the error (ledger id, entry id etc). This would simplify the loops in the test itself, as you wouldn't have to do anything to get the ledger and entry id, as it'd already be in the exception message. In fact, the loop could be reduced to

```executor.invokeAll(writeTasks, 6, TimeUnit.SECONDS).forEach(f -> f.join())```

Reply:
invokeAll returns list of Futures and join is method of CompletableFuture but not Future.

Reply:
other than removing redundant .isDone check, I would keep the rest of the test logic as it is now. For diagnosing in the case of failure, it is helpful to have assert failures and proper error log lines (by catching the error and log it) rather than failing with generic cancellation exception or executionexception when we call future.get().

Reply:
It was for diagnosing the errors that I suggested putting the ledgerid and entryid in the message of the thrown exception. You're right about join though. Using get would make things messy, so possibly not worth it.

Reply:
I don't see why using .get would be messy. Just FYI about Junit, "The JUnit framework captures only assertion errors in the main thread running the test. It is not aware of exceptions from within new spawn threads. In order to do it right, you should communicate the thread's termination state to the main thread." So to get the status/result of task executed in non-main thread we should call for .get of future. Agreed, instead of returning boolean we may throw exception with proper message, but for diagnosing / logging purpose I feel it is better to fail with Assert statements  and enough logging in the tasks. I see it is just a matter of preference and there is no much difference in choosing one over the other.

Reply:
because .get() throws checked exceptions, so in the snippet I posted, you would have to catch and rethrow as an unchecked exception. 

I was suggest throwing exceptions in the callable thread precisely because asserts are only meaningful in the main thread (they're java.lang.Error). Throwing an exception in the throwable will actually be easer to debug, because you'll get the stacktrace directly on failure, so you can tell what line failed without having to even go to the logs.

Reply:
first thing 

1) calling .get() on future is inevitable, since thats the only way you would get the result of the future.

> Instead of returning false on errors, you could throw Exception, and have that carry the context of the error (ledger id, entry id etc). This would simplify the loops in the test itself.

2) even if I change the logic to throw exception instead of boolean in the case of failure, we are not getting rid of loops. Because incase of CancellationException we would get generic cancellation exception instead of any of our exception with customized message. So atleast for this case we need to do what we are doing.

Again as I said this is matter of preference how we deal with return status of tasks/futures in junit test env. If you see it as a dealbreaker for you, I'll go ahead and change it to throw exception, as I don't see value in dragging this further.